### DISCUSSION TITLE
it is possible to make excel call functions asynchronously?

### DISCUSSION THREAD

**Discussion by SKYC829** [Excel-DNA/ExcelDna#565] [Q&A] [answered]
it is possible to make excel call functions asynchronously? 
just like .net core web api pipline how  works, make excel row by rows calculation, parallel calculation cell in current row or make excel column by columns calculation but parallel calculation per cell in current column

**Comment by govert** [comment, marked as answer]
Yes - you can make asynchronous functions with Excel-DNA.
Best is to make your function as a normal .NET async function that returns Task<T>. To glue this up with Excel you have a few options:
* Use the Registration extension library - see https://github.com/Excel-DNA/Registration
* Just pull in two helper files from the Registration extensions into your project, and call from there. See https://excel-dna.github.io/TestDocs/docs/guides-basic/asynchronous-functions and here: https://github.com/Excel-DNA/Samples/tree/master/AsyncHelloWorld
* The Task-base support builds on the Excel RTD mechanism, and you can also use this directly.
* Finally, there are also "native" async functions in Excel. These are a bit less friendly to the user than the above options, so I suggest you avoid these for a start.

**Reply by SKYC829** [reply]
> Yes - you can make asynchronous functions with Excel-DNA. Best is to make your function as a normal .NET async function that returns Task. To glue this up with Excel you have a few options:
> 
>     * Use the Registration extension library - see https://github.com/Excel-DNA/Registration
> 
>     * Just pull in two helper files from the Registration extensions into your project, and call from there. See https://excel-dna.github.io/TestDocs/docs/guides-basic/asynchronous-functions and here: https://github.com/Excel-DNA/Samples/tree/master/AsyncHelloWorld
> 
>     * The Task-base support builds on the Excel RTD mechanism, and you can also use this directly.
> 
>     * Finally, there are also "native" async functions in Excel. These are a bit less friendly to the user than the above options, so I suggest you avoid these for a start.

hi govert,
i have some new question, this argument returns error reference in here cause function wizard dose not show calculation result.
but it can transfer a correct value in final logic, how to make it show the correct value in function wizard?
![screenshot_20230208113455](https://user-images.githubusercontent.com/60951390/217441178-35c0f863-9b46-4fe9-b34c-05ac912189eb.png)

**Comment by govert** [comment]
I don't think you can get the function wizard to work for async functions like that. The only way might be like this, to check when you are in the function wizard, and then to call the Task-returning function synchronously. But the user experience might not be great if your function is slow, as it will wait for the calculation result every time the change an argument in the dialog.

```C#
public static object MyFunction(object arg1, object arg2)
{
    if (ExcelDnaUtil.IsInFunctionWizard())
    {
        var functionTask = MyFunctionImplAsync(arg1, arg2);
        return functionTask.Result;
    }
    else
    {
        var functionName = nameof(MyFunction);
        var parameters = new object[] { arg1, arg2};
        return AsyncTaskUtil.RunTask(functionName, parameters, () => MyFunctionImplAsync( arg1, arg2 ));
    }
}
```

**Reply by SKYC829** [reply]
> I don't think you can get the function wizard to work for async functions like that. The only way might be like this, to check when you are in the function wizard, and then to call the Task-returning function synchronously. But the user experience might not be great if your function is slow, as it will wait for the calculation result every time the change an argument in the dialog.
> 
> ```cs
> public static object MyFunction(object arg1, object arg2)
> {
>     if (ExcelDnaUtil.IsInFunctionWizard())
>     {
>         var functionTask = MyFunctionImplAsync(arg1, arg2);
>         return functionTask.Result;
>     }
>     else
>     {
>         var functionName = nameof(MyFunction);
>         var parameters = new object[] { arg1, arg2};
>         return AsyncTaskUtil.RunTask(functionName, parameters, () => MyFunctionImplAsync( arg1, arg2 ));
>     }
> }
> ```

Alright. thank you very much. :)
